### ISSUE TITLE
keyborad shortcut for folding/unfolding of headings

### ISSUE THREAD

**Issue by alexysong** [Zettlr/Zettlr#857]
## Description
Add keyboard shortcuts for the folding of headings. My suggested key bindings are:

1. `Ctrl + ;` (fold and unfold)
2. `Ctrl + Tab` (both fold and unfold)
3. `Ctrl + [` (fold), `Ctrl + ]` (unfold). 

To have fold/unfold to be separate key bindings may seem logical at a first glance. However, to have one key binding for both actually makes sense. A folded heading can't do anything else but to be unfolded, and vice versa. Also, to have one key binding for both saves.

## Example Workflow
When my cursor is on a heading, press a key to fold/unfold at this heading.
So I don't have to lift my hand away from the keyboard and use my mouse to click on the little triangle to do this. 

## Caveats
Not that I can think of.

## Additional Information
Or, allow user-defined keyboard shortcuts?
Also, if you search "fold" in the doc : [fold](https://docs.zettlr.com/en/search.html?q=fold), no useful info regarding folding on headings can be found.

**Comment by nathanlesage** [comment]
On suggesting new keyboard shortcuts, it's tradition that you have the obligation of finding one that suits, by testing that it's not being used (helpful hint: https://www.zettlr.com/shortcuts) + making sure other programs have something similar.

> Also, if you search "fold" in the doc : fold, no useful info regarding folding on headings can be found.

Feel free to [improve the documentation](https://github.com/zettlr/zettlr-docs)!

**Comment by alexsong19** [comment]
I've updated it by added a few suggested key bindings.

**Comment by derwok** [comment]
I also would like to fold/unfold via keyboard.
But I think the proposed keys might cause problems in multiple countries where `[`and `;` are not easy to reach... or only work with `Shift` or so...
Additionally: `Ctrl+Tab` is common key for navigating through open Tabs of browsers / editors.

**Comment by caseywatts** [comment]
I'm looking for this feature too! I love keyboard shortcuts, and I use folds a lot.


## toggle
As another reference implementation, I use vim (or atom/vscode in vim mode): vim relies on `z` for folds since it looks like a folded line. I could imagine something like `cmd/ctrl + alt + z` as another option, although the ones above sound good to me, too.

## non-toggle
I really like having the "fold" and "unfold" non-toggle option too - sometimes I use it to fold wherever I am, the parent foldable thing `Ctrl + [ (fold), Ctrl + ] (unfold).` sounds natural to me for that.

Implementation, this may be a little more work to implement than toggling for a line you're already on, since you have to go up to find any parent element to fold. We could make this a separate feature request if that makes sense, np.

**Comment by nathanlesage** [comment]
You know guys? I'm blind: Zettlr supports that keyboard shortcut since we included the sublime commands 😂

Here's a snippet from the keys Sublime defines:

```javascript
    "Shift-Ctrl-[": "fold",
    "Shift-Ctrl-]": "unfold",
```

However, this won't work on German layouts as you'd have to press <kbd>Option</kbd> (macOS) or <kbd>AltGr</kbd> (Windows) for the `[` and `]` characters, and then, curiously, these events don't trigger. For this, we'd have to wait for PR #927

One thing however: I updated the foldcode plugin to also attempt to find a range even if there is none on the current line (allows you to fold Markdown headings from within their section)

**Comment by caseywatts** [comment]
that sounds great! love leveraging their existing commands :)

FYI even on the line of the fold, it's not working for me right now, there could be a wider issue, not just with your keyboard layouts. I'm on Zettlr version 1.7.1 Windows, using keyboard layouts US Dvorak and US English. No problem waiting for that other PR, just sharing in case it helps debug.

**Comment by annargrs** [comment]
"shift-control-[" and "shift-control-[" do not work for me either. Ubuntu 18.04, Zettlr 1.7.1, US English layout.

**Comment by nathanlesage** [comment]
Oh, damn it …

**Comment by karstenpe** [comment]
I am currently using boostnote where I have added two shortcuts for foldAll and unfoldAll; first by patching the keymap file (sublime.js) and later via configuring custom shortcuts. Looking into the code (of both, boostnote and zettlr) it looks like many components are the same (codemirror plus fold addon).
There are also existing mappings in the code for folding. Interestingly there are some mappings obviously by holding CTRL and pressing two keys in sequence ( Ctrl-K Ctrl-1": "foldAll",  "Ctrl-K Ctrl-0": "unfoldAll",   "Ctrl-K Ctrl-J": "unfoldAll",  "Ctrl-Alt-Up": "addCursorToPrevLine", etc.).
It looks to me as if zettlr is "cut off" from these mappings by using (simpel) CTRL+K for another functionality.
Maybe this points to a simpler way leveraging existing mapping!?

**Comment by caseywatts** [comment]
It looks like ctrl+k is being used for links - and I actually do use that a lot now that I'm thinking about it! That's the keybinding other apps like Gmail use for links.
https://github.com/jeffgeorge/Zettlr/blob/master/source/renderer/assets/codemirror/zettlr-plugin-markdown-shortcuts.js#L593-L594

I wonder what other apps do for foldAll and unfoldAll? I'd love to be able to do those with the keyboard too.

**Comment by nathanlesage** [comment]
Feel free to propose keyboard shortcuts for these commands:

```javascript
CodeMirror.commands.toggleFold = function(cm) {
    cm.foldCode(cm.getCursor());
  };
  CodeMirror.commands.fold = function(cm) {
    cm.foldCode(cm.getCursor(), null, "fold");
  };
  CodeMirror.commands.unfold = function(cm) {
    cm.foldCode(cm.getCursor(), null, "unfold");
  };
  CodeMirror.commands.foldAll = function(cm) {
    cm.operation(function() {
      for (var i = cm.firstLine(), e = cm.lastLine(); i <= e; i++)
        cm.foldCode(CodeMirror.Pos(i, 0), null, "fold");
    });
  };
  CodeMirror.commands.unfoldAll = function(cm) {
    cm.operation(function() {
      for (var i = cm.firstLine(), e = cm.lastLine(); i <= e; i++)
        cm.foldCode(CodeMirror.Pos(i, 0), null, "unfold");
    });
```

Fold and unfold should already be defined, but I'm not sure anymore which keyboard shortcut 🙄

**Comment by karstenpe** [comment]
> It looks like ctrl+k is being used for links - and I actually do use that a lot now that I'm thinking about it! That's the keybinding other apps like Gmail use for links.
> https://github.com/jeffgeorge/Zettlr/blob/master/source/renderer/assets/codemirror/zettlr-plugin-markdown-shortcuts.js#L593-L594
> 
> I wonder what other apps do for foldAll and unfoldAll? I'd love to be able to do those with the keyboard too.

Yes, CTRL+K is quite common for links; I would not change this, but, as I mentioned:
>> there are some mappings obviously by holding CTRL and pressing **two keys** in sequence
I wonder if this could be made available; would  be a starting point for us fans of fast folding without interference.

Will have a look into the snippet @nathanlesage posted to understand better.

**Comment by nathanlesage** [comment]
@karstenpe Have a look at the file "generate-keymap.js" as well, as there new keybindings are defined.

**Comment by karstenpe** [comment]
thanks a lot!
Playing around already a little with:
keymap['Alt-1'] = 'fold'      
keymap['Alt-2'] = 'unfold'      
keymap['Alt-9'] = "unfoldAll",
keymap['Alt-0'] = 'foldAll'    
Not sure about the **two key** sequences mentioned above. But, no problem. I will investigate that later.

**Comment by pcarmich** [comment]
Could you post a short description of how you define custom shortcuts? The keymap['Alt-1'] = 'fold'and keymap['Alt-2'] = 'unfold' did not work for me on Windows, zettlr 1.8.2. Nor did the 'two key' sequences mentioned above.  I'm a newbie, so... not sure from the documentation if these options are exposed through a config file to the user? It would be a nice feature if it was.  Also, I have a small opinion on UX.  Coming from emacs-org mode, I find using the Tab key for cycling through different folding states to be a very elegant solution.

**Comment by nathanlesage** [comment]
I would advise against using two-key strokes, because currently there's no indicator for those (as, e.g., in VSCode). That might confuse the main target audience (non-tech users). But I may just have an idea of how to manage custom shortcuts on the editor … this way we could circumvent all those questions!

**Comment by SolarDesalination** [comment]
Fold unfold would be very useful. Upvoting this request.

**Comment by bmalbusca** [comment]
Upvoting this request!

**Comment by Etienne-TR** [comment]
These shortcuts will be very useful to me to navigate into a note !

I don't get all that was said here. Is there already a solution to make cutom shortcuts to toggle sections ?

I try to find informations but I find nothing in the docs : https://docs.zettlr.com/en/reference/shortcuts/  
And nothing I can't use by myself in the repo : #818

**Comment by Etienne-TR** [comment]
> I would advise against using two-key strokes, because currently there's no indicator for those (as, e.g., in VSCode). That might confuse the main target audience (non-tech users). But I may just have an idea of how to manage custom shortcuts on the editor … this way we could circumvent all those questions!

It is a big cliffhanger to me ! ;-) How custom shortcuts could be managed ?

**Comment by nathanlesage** [comment]
> It is a big cliffhanger to me ! ;-) How custom shortcuts could be managed ?

Well, once I got rid of most of the biggest bugs, I already have a plan: Find a small but reliable package to manage shortcuts and then just capture everything in the renderer. It's a somewhat more complicated feature (if I want it to be useful and simple), so I won't promise anything for 3.0.0, but maybe 3.1!

**Comment by Etienne-TR** [comment]
I have a look here : https://codemirror.net/docs/ref/#commands

If found these shortcuts : 

```
[foldKeymap](https://codemirror.net/docs/ref/#language.foldKeymap): readonly [KeyBinding](https://codemirror.net/docs/ref/#view.KeyBinding)[]

    Default fold-related key bindings.

        Ctrl-Shift-[ (Cmd-Alt-[ on macOS): [foldCode](https://codemirror.net/docs/ref/#language.foldCode).
        Ctrl-Shift-] (Cmd-Alt-] on macOS): [unfoldCode](https://codemirror.net/docs/ref/#language.unfoldCode).
        Ctrl-Alt-[: [foldAll](https://codemirror.net/docs/ref/#language.foldAll).
        Ctrl-Alt-]: [unfoldAll](https://codemirror.net/docs/ref/#language.unfoldAll).
```
I can't use them.

Is that one of the reason custom shortcuts are needed ? Or there is a way to use the default key-bindings ?

**Comment by nathanlesage** [comment]
Default keybindings are basically one line of code, so I can add them! And hope they don't interfere :D

**Comment by notuntoward** [comment]
Agreed.  A single key to toggle the folding state would be very useful, and is common in other outliner-type programs.  In fact the current obsidian header folding method is already a toggle, but it's only available by a clickable symbol.

**The big problem**: there's no way to make a hotkey for a header folding toggle.

**Comment by lewallen** [comment]
I hope this thread is an appropriate place for this post, but: in case it's useful to anyone, I made a fork here https://github.com/lewallen/Zettlr-Sam.git which has keybindings for the fold commands, if you build and package that fork as per the instructions in the README, it should work as expected with with the addition of the folding commands (to get different keybindings just modify the new code, as described below). FWIW the modifications are all in `/source/common/modules/markdown-editor/commands/keymap.ts` and look like

```ts
/* add this to the import list at the top of the file */
import { foldAll, unfoldAll, unfoldCode, foldCode, toggleFold } from '@codemirror/language'
```

```ts
/* add this to the top of the "KeyBinding" list */
export const customKeymap: KeyBinding[] = [
    { key: 'Mod-[', run: foldCode },
    { key: 'Mod-]', run: unfoldCode },
    { key: 'Mod-Shift-,', run: foldAll },
    { key: 'Mod-Shift-.', run: unfoldAll },
    ... /* plus the rest of the keybindings that are already there */
```

I don't know much about contributing/pulling on github, and the above keybindings are anyway just my personal preference, but if that's something to consider just let me know.
